vm/state_tree/src/lib.rs

```diff
@@ -173,30 +327,49 @@ where
         Ok(new_addr)
     }
 
-    // TODO update snapshotting to not flush tree: https://github.com/ChainSafe/forest/issues/373
-    /// Persist changes to store and return Cid to revert state to.
-    pub fn snapshot(&mut self) -> Result<Cid, String> {
-        self.flush()
+    /// Add snapshot layer to stack.
+    pub fn snapshot(&mut self) -> Result<(), Box<dyn StdError>> {
+        self.snaps.add_layer();
+        Ok(())
     }
 
-    /// Revert to Cid returned from `snapshot`
-    pub fn revert_to_snapshot(&mut self, cid: &Cid) -> Result<(), String> {
-        // Update Hamt root to snapshot Cid
-        self.hamt.set_root(cid).map_err(|e| e.to_string())?;
+    /// Merges last two snap shot layers
+    pub fn clear_snapshot(&mut self) -> Result<(), Box<dyn StdError>> {
+        Ok(self.snaps.merge_last_layer()?)
+    }
 
-        self.actor_cache = Default::default();
+    /// Revert state cache by removing last snapshot
+    pub fn revert_to_snapshot(&mut self) -> Result<(), Box<dyn StdError>> {
+        self.snaps.drop_layer()?;
+        self.snaps.add_layer();
         Ok(())
     }
 
     /// Flush state tree and return Cid root.
-    pub fn flush(&mut self) -> Result<Cid, String> {
-        for (addr, act) in self.actor_cache.read().iter() {
-            // Set each value from cache into hamt
-            self.hamt
-                .set(addr.to_bytes().into(), act.clone())
-                .map_err(|e| e.to_string())?;
+    pub fn flush(&mut self) -> Result<Cid, Box<dyn StdError>> {
+        if self.snaps.layers.len() != 1 {
+            return Err(format!(
+                "tried to flush state tree with snapshots on the stack: {:?}",
+                self.snaps.layers.len()
+            )
+            .into());
+        }
+
+        for (addr, sto) in self.snaps.layers[0].actors.read().iter() {
+            match sto {
+                None => {
+                    self.hamt
+                        .delete(&addr.to_bytes())
+                        .map_err(|e| e.to_string())?;
+                }
+                Some(ref state) => {
+                    self.hamt
+                        .set(addr.to_bytes().into(), state.clone())
+                        .map_err(|e| e.to_string())?;
+                }
+            }
         }
 
-        self.hamt.flush().map_err(|e| e.to_string())
+        Ok(self.hamt.flush().map_err(|e| e.to_string())?)
```

Comment:
All of these `to_string()` calls defeat the purpose of the dynamic errors, you should remove all of these